### PR TITLE
some refactor in account module and implement file module

### PR DESCRIPTION
**Pull request by MiladRezaeian**

### CODE REVIEW

File: modules/Account/app/Models/User.php [hunk 1]
```diff
@@ -77,4 +85,10 @@ public static function factory()
     {
         return UserFactory::class;
     }
+
+    public function ownedFiles()
```
Comment (yoyodeveloper20) [inline modules/Account/app/Models/User.php:89]: The method name must be files

File: modules/Account/app/Contracts/Services/UserServiceInterface.php [hunk 2]
```diff
@@ -2,6 +2,24 @@
...
+
+    public function store(RegisterRequest $request);
+
+    public function show(User $user);
+
+    public function update(RegisterRequest $request, User $user);
+
+    public function destroy(User $user);
+
+    public function search(Request $request);
```
Comment (yoyodeveloper20) [inline modules/Account/app/Contracts/Services/UserServiceInterface.php]: remove this method

File: modules/Account/app/Http/Controllers/Api/Web/Panel/V1/UserController.php [hunk 3]
```diff
@@ -7,22 +7,25 @@
 use Illuminate\Http\Response;
 use Modules\Account\app\Contracts\Services\UserServiceInterface;
 use Modules\Account\app\Http\Requests\Auth\RegisterRequest;
```
Comment (yoyodeveloper20) [inline modules/Account/app/Http/Controllers/Api/Web/Panel/V1/UserController.php]: change app to App

File: modules/Account/app/Repositories/PermissionRepository.php [hunk 4]
```diff
@@ -10,9 +10,17 @@
...
+
+    public function __construct(Permission $permission)
+    {
+        $this->permission = $permission;
+    }
+
+    public function getAllWithRolesAndSearchCriteria($params)
     {
-        $permissions = Permission::paginate();
+        $permissions = $this->applySearchFilters($params);
```
Comment (yoyodeveloper20) [inline modules/Account/app/Repositories/PermissionRepository.php]: remove applySearchFilters method

File: modules/Account/app/Repositories/RoleRepository.php [hunk 5]
```diff
@@ -18,9 +18,17 @@
...
+
+    public function __construct(Role $role)
+    {
+        $this->role = $role;
+    }
+
+    public function getAllWithPermissionsAndSearchCriteria($params)
     {
-        $roles = Role::paginate();
+        $roles = $this->applySearchFilters($params);
```
Comment (yoyodeveloper20) [inline modules/Account/app/Repositories/RoleRepository.php]: remove applySearchFilters method

File: modules/Account/app/Repositories/PermissionRepository.php [hunk 6]
```diff
@@ -48,4 +56,23 @@ public function destroy(Permission $permission)
...
+
+        if (!empty($name)) {
+            $query->where('name', 'like', '%' . $params['name'] . '%');
+        }
+
+        if (!empty($translated_name)) {
+            $query->where('translated_name', 'like', '%' . $params['translated_name'] . '%');
+        }
+
+        return $query->paginate();
```
Comment (yoyodeveloper20) [inline modules/Account/app/Repositories/PermissionRepository.php]: check per_page and page
set max for per_page

File: modules/BaseModuleProvider.php [hunk 7]
```diff
@@ -38,16 +35,13 @@ public function boot()
 //        app(Factory::class)->load($moduleBasePath . '/database/factories');
     }
 
-    private function defineRoute(string $moduleBasePath): void
-    {
-        $this->loadRoutesFrom($moduleBasePath . '/routes/api/web/customer/v1.php');
-        $this->loadRoutesFrom($moduleBasePath . '/routes/api/web/guest/v1.php');
-        $this->loadRoutesFrom($moduleBasePath . '/routes/api/web/panel/v1.php');
-    }
+    protected function defineRoute(string $moduleBasePath): void {}
```
Comment (yoyodeveloper20) [inline modules/BaseModuleProvider.php]: it's not okey

File: modules/File/app/Constants/FileConstants.php [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+class FileConstants
+{
+
+    const FILE_TYPE_DEFAULT = 'D';
+    const FILE_TYPE_NATIONAL_CARD = 'NC';
+    const FILE_TYPE_PROFILE = 'P';
+    const FILE_TYPE_PRODUCT_IMAGE = 'PI';
+
+    const FILE_TYPES_DETAILS = [
+        self::FILE_TYPE_DEFAULT => [
```
Comment (yoyodeveloper20) [inline modules/File/app/Constants/FileConstants.php:14]: add max size

File: modules/File/app/Http/Controllers/Api/Web/Panel/V1/FileController.php [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace Modules\File\app\Http\Controllers\Api\Web\Panel\V1;
+
+use App\Http\Controllers\BaseController;
+use Illuminate\Http\Response;
+use Modules\File\app\Contracts\Services\FileServiceInterface;
+use Modules\File\app\Http\Requests\Api\Web\Panel\V1\FileRequest;
+use Modules\File\app\Models\File;
+
+class FileController extends BaseController
```
Comment (yoyodeveloper20) [inline modules/File/app/Http/Controllers/Api/Web/Panel/V1/FileController.php]: remove this controller

File: modules/File/app/Http/Requests/Api/Web/Panel/V1/FileRequest.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array<string, \Illuminate\Contracts\Validation\ValidationRule|array<mixed>|string>
+     */
+    public function rules(): array
+    {
+        return [
+            'type' => ['required', 'string'],
+            'file' => ['required', 'file', 'mimetypes:image/jpeg,image/png', 'max:1024'],
```
Comment (yoyodeveloper20) [inline modules/File/app/Http/Requests/Api/Web/Panel/V1/FileRequest.php]: update with config

File: modules/File/app/Models/File.php [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+{
+
+    use HasFactory,
+        SoftDeletes;
+
+    protected $fillable = [
+        'name', 'description', 'type', 'path', 'is_public', 'verified', 'size', 'mime_type', 'fileable_type', 'fileable_id','user_id'
+    ];
+
+    public function absolutePath()
```
Comment (yoyodeveloper20) [inline modules/File/app/Models/File.php]: move this method to FileService class

File: modules/File/app/Models/File.php [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        return resolve(StorageManagerService::class)->getAbsolutePathOf($this->name . $this->type, $this->is_public);
+    }
+
+    public function fileable(): MorphTo
+    {
+        return $this->morphTo();
+    }
+
+//    public function user()
```
Comment (yoyodeveloper20) [inline modules/File/app/Models/File.php]: remove this comments

File: modules/File/app/Providers/FileModuleProvider.php [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+        App::bind(FileServiceInterface::class, FileService::class);
+        App::bind(StorageManagerInterface::class, StorageManagerService::class);
+    }
+
+    private function registerRepositoryBindings()
+    {
+        App::bind(FileRepositoryInterface::class, FileRepository::class);
+    }
+
+    private function defineModelBindings()
```
Comment (yoyodeveloper20) [inline modules/File/app/Providers/FileModuleProvider.php]: empty method is bad smell

File: modules/File/app/Providers/FileModuleProvider.php [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    {
+
+    }
+
+    private function definePermissions()
```
Comment (yoyodeveloper20) [inline modules/File/app/Providers/FileModuleProvider.php]: empty method is bad smell

File: modules/File/app/Repositories/FileRepository.php [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        return $file->delete();
+    }
+
+    public function forceDelete(File $file)
+    {
+        return $file->forceDelete();
+    }
+
+    public function isFileRowExists($user_id, $type)
```
Comment (yoyodeveloper20) [inline modules/File/app/Repositories/FileRepository.php]: search by fileable_id , fileable_type and type

File: modules/File/app/Services/FileService.php [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+    private StorageManagerInterface $storageManagerService;
+    public function __construct(FileRepositoryInterface $fileRepository, StorageManagerService $storageManagerService)
+    {
+        $this->fileRepository = $fileRepository;
+        $this->storageManagerService = $storageManagerService;
+    }
+
+    public function store($model, $file, $type, $path = null, $description = null)
+    {
+        $user_id = auth()->id();
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/FileService.php]: camelCase

File: modules/File/app/Services/FileService.php [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+            'is_public' => $isPublic,
+            'verified' => $verified,
+            'size' => $size,
+            'mime_type' => $mime_type,
+            'fileable_type' => $fileable_type,
+            'fileable_id' => $fileable_id,
+            'user_id' => $user_id,
+        ];
+
+        if ($this->isReplaceable($type)) {
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/FileService.php]: clean code for if and else

File: modules/File/app/Services/FileService.php [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+
+            try {
+
+                $existsFileRow = $this->fileRepository->isFileRowExists($user_id, $type);
+
+                DB::transaction(function () use ($isPublic, $fileName, $data, $existsFileRow) {
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/FileService.php]: edit usage of transaction

File: modules/BaseModuleProvider.php (hunk 7 above)
Comment (MiladRezaeian) [reply modules/BaseModuleProvider.php]: I do not know how to resolve

File: modules/File/app/Services/StorageManagerService.php [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+use Illuminate\Http\UploadedFile;
+use Illuminate\Support\Facades\Storage;
+use Modules\File\App\Contracts\Services\StorageManagerInterface;
+
+class StorageManagerService implements StorageManagerInterface
+{
+    public function putFile($fileName, UploadedFile $file, string $path, $isPublic = false)
+    {
+        $savedPath = $this->getDisk($isPublic)->putFileAs($path, $file, $fileName);
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/StorageManagerService.php]: setDisk method:
$this->disk = $isPublic ? Storage::disk('public') : Storage::disk('private');
return $this;

File: modules/File/app/Services/StorageManagerService.php [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+
+    public function deleteFile(string $path, bool $isPublic)
+    {
+        $relativePath = str_replace('private/', '', $path); //TODO
+        return $this->getDisk($isPublic)->delete($relativePath);
+    }
+
+    private function getDisk(bool $isPublic)
+    {
+        return $isPublic ? Storage::disk('public') : Storage::disk('private');
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/StorageManagerService.php]: return $this->disk;

File: modules/File/app/Services/StorageManagerService.php [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+class StorageManagerService implements StorageManagerInterface
+{
+    public function putFile($fileName, UploadedFile $file, string $path, $isPublic = false)
+    {
+        $savedPath = $this->getDisk($isPublic)->putFileAs($path, $file, $fileName);
+        $fullPath = $this->getDisk($isPublic)->url($savedPath);
+
+        $baseUrl = config('app.url');
+        $relativePath = str_replace($baseUrl . '/storage/', '', $fullPath); // Remove the "/storage" prefix //TODO
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/StorageManagerService.php]: check link in skype

File: modules/File/app/Services/StorageManagerService.php [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+
+        return $relativePath;
+    }
+
+    public function isFileExists(string $path, bool $isPublic)
+    {
+        $relativePath = str_replace('private/', '', $path); //TODO
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/StorageManagerService.php]: 'private/' in hard code

File: modules/File/app/Services/StorageManagerService.php [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+        return $this->getDisk($isPublic)->exists($relativePath);
+    }
+
+    public function deleteFile(string $path, bool $isPublic)
+    {
+        $relativePath = str_replace('private/', '', $path); //TODO
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/StorageManagerService.php]: remove hard code

File: modules/File/app/Services/FileService.php [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+        $this->fileRepository = $fileRepository;
+        $this->storageManagerService = $storageManagerService;
+    }
+
+    public function store($model, $file, $type, $path = null, $description = null)
+    {
+        $fileableType = get_class($model);
+        $fileableId = $model->id;
+        $userId = auth()->id();
+        $fileName = $fileableId . '_' . $file->getClientOriginalName();
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/FileService.php]: random char for prevent duplicate name

File: modules/File/app/Services/FileService.php [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+        $isPublic = $this->checkTypeIsPublic($type);
+        $verified = 1;
+        $size = $file->getSize();
+        $mime_type = $file->getClientMimeType();
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/FileService.php]: camleCase for $mimeType

File: modules/File/app/Services/FileService.php [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+        if (!$this->isReplaceable($type)) {
+            return $this->fileRepository->create($data);
+        }
+
+        try {
+            $existsFileRow = $this->fileRepository->isFileRowExists($fileableId, $fileableType, $type);
+
+            DB::beginTransaction();
+
+            if ($existsFileRow instanceof File) {
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/FileService.php]: $existFileRow

File: modules/File/app/Services/FileService.php [hunk 27]
```diff
@@ -0,0 +1,107 @@
...
+                $this->fileRepository->forceDelete($existsFileRow);
+
+                if ($this->storageManagerService->isFileExists($existsFileRow->path, $existsFileRow->is_public) && $fileName != $existsFileRow->name) {
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/FileService.php]: is fileExistPhisically
write method and that method must get fileId or fileObject and ...

File: modules/File/app/Services/FileService.php [hunk 28]
```diff
@@ -0,0 +1,107 @@
...
+                    $this->storageManagerService->deleteFile($existsFileRow->path, $existsFileRow->is_public); //TODO issue of deleting exists file that have uploaded
```
Comment (yoyodeveloper20) [inline modules/File/app/Services/FileService.php]: refactor deleteFile base on FileId or FileObject

**Pull request merged**